lib/auth/auth_with_roles_test.go

```diff
@@ -128,118 +127,151 @@ func TestGenerateDatabaseCert(t *testing.T) {
-		require.NoError(t, err)
-		require.Empty(t, resourceDiff(storedAuthPref, dynamicAuthPref))
-	})
-
-	newDynamicAuthPref, err := types.NewAuthPreference(types.AuthPreferenceSpecV2{
-		SecondFactor: constants.SecondFactorOTP,
-	})
-	require.NoError(t, err)
-	t.Run("from dynamic to another dynamic", func(t *testing.T) {
-		err = server.SetAuthPreference(ctx, newDynamicAuthPref)
-		require.NoError(t, err)
-		storedAuthPref, err = server.GetAuthPreference(ctx)
-		require.NoError(t, err)
-		require.Empty(t, resourceDiff(storedAuthPref, newDynamicAuthPref))
-	})
-
-	staticAuthPref := newU2FAuthPreferenceFromConfigFile(t)
-	t.Run("from dynamic to static", func(t *testing.T) {
-		err = server.SetAuthPreference(ctx, staticAuthPref)
-		require.NoError(t, err)
-		storedAuthPref, err = server.GetAuthPreference(ctx)
-		require.NoError(t, err)
-		require.Empty(t, resourceDiff(storedAuthPref, staticAuthPref))
-	})
-
-	newAuthPref, err := types.NewAuthPreferenceFromConfigFile(types.AuthPreferenceSpecV2{
-		SecondFactor: constants.SecondFactorOTP,
-	})
-	require.NoError(t, err)
-	replaceStatic := func(success bool) func(t *testing.T) {
+	testOperation := func(op func(*ServerWithRoles) error, allowRules []types.Rule, expectErr, withConfigFile bool) func(*testing.T) {
 		return func(t *testing.T) {
-			err = server.SetAuthPreference(ctx, newAuthPref)
-			checkSetResult := require.Error
-			if success {
-				checkSetResult = require.NoError
+			if withConfigFile {
+				p.storeConfigFile(testAuth.AuthServer)
+			} else {
+				p.storeDefault(testAuth.AuthServer)
 			}
-			checkSetResult(t, err)
-
-			storedAuthPref, err = server.GetAuthPreference(ctx)
-			require.NoError(t, err)
-			expectedStored := staticAuthPref
-			if success {
-				expectedStored = newAuthPref
+			server := serverWithAllowRules(t, testAuth, allowRules)
+			opErr := op(server)
+			if expectErr {
+				require.Error(t, opErr)
+			} else {
+				require.NoError(t, opErr)
 			}
-			require.Empty(t, resourceDiff(storedAuthPref, expectedStored))
 		}
 	}
 
-	t.Run("replacing static fails without VerbCreate privilege", replaceStatic(false))
-
-	// Grant VerbCreate privilege for cluster_auth_preference.
-	allowRules[0].Verbs = append(allowRules[0].Verbs, types.VerbCreate)
-	server = withAllowRules(t, testAuth, allowRules)
+	// runTestCases generates all non-empty RBAC verb combinations and checks the expected
+	// error for each operation.
+	runTestCases := func(withConfigFile bool) {
+		for i := 1; i < 8; i++ {
+			verbs := []string{}
+			expectGetErr, expectSetErr, expectResetErr := true, true, true
+			if (i&1) == 1 || p.alwaysReadable {
+				verbs = append(verbs, types.VerbRead)
+				expectGetErr = false
+			}
+			if (i & 2) == 2 {
+				verbs = append(verbs, types.VerbUpdate)
+				if !withConfigFile {
+					expectSetErr, expectResetErr = false, false
+				}
+			}
+			if (i & 4) == 4 {
+				verbs = append(verbs, types.VerbCreate)
+			}
+			if (i & 6) == 6 {
+				expectSetErr = false
+			}
```

Comment:
nit: I think this makes the intention more clear, feel free to ignore
```suggestion
			if (i & 4) == 4 {
				verbs = append(verbs, types.VerbCreate)
				if (i & 2) == 2 {
					expectSetErr = false
				}
			}
```